src/OrleansRuntime/Core/IInvokable.cs

```diff
@@ -0,0 +1,12 @@
+using Orleans.CodeGeneration;
+
+namespace Orleans.Runtime
+{
+    /// <summary>
+    /// Common internal interface for SystemTarget and ActivationData.
+    /// </summary>
+    internal interface IInvokable
+    {
+        IGrainMethodInvoker GetInvoker(GrainTypeManager typeManager, int interfaceId, string genericGrainType = null);
```

Comment:
I wonder if it makes sense to pass another `IInvokable` instead of `GrainTypeManager` and make it implement IInvokable. This way you can build a strategy chain instead of forcing a single decorated object of type GrainTypeManager. Just a naive suggestion, and don't know if it even makes sense logically.